### ISSUE TITLE
Understanding optimization process with ColBERT

### ISSUE THREAD

**Issue by puzzlecollector** [stanford-futuredata/ColBERT#100]
@okhat 

From what I understand, with ColBERT, this is how the model is trained: 

query -> BERT -> query embedding (I will call this vector q) 
positive document -> BERT -> positive embedding (I will call this vector p) 
negative document -> BERT -> negative embedding (I will call this vector n) 

if S is the metric used for scoring (e.g. cosine similarity)

s1 = S(q, p) 
s2 = S(q, n) 

Then we apply softmax on [s1, s2] and get probability return value, for example [0.6, 0.4]. Now we compare this with the true label, which is always 0 and optimize BERT using binary cross entropy. 

1. Is this understanding of mine correct? 
2. What really is the effect of setting the true label to be zero all the time? How is the network correctly optimized this way such that it ranks relevant documents effectively?

**Comment by puzzlecollector** [comment]
@okhat 
Regarding my question 2 above, I guess it's because by maximizing the score for label 0, we maximize S(q, p), whereas we minimize S(q, n) right? this way the similarity between the query and the positive document will have a greater value, whereas the similarity value between positive and negative will be smaller.

**Comment by okhat** [comment]
Your understanding is correct. The only remark is that q, p, and n in your notation are matrices, not vectors.
